src/Nethermind/Nethermind.Merge.Plugin/Handlers/V1/GetPayloadBodiesByHashV1Handler.cs

```diff
@@ -12,30 +12,28 @@
 
 namespace Nethermind.Merge.Plugin.Handlers.V1
 {
-    public class GetPayloadBodiesV1Handler : IAsyncHandler<Keccak[], ExecutionPayloadBodyV1Result[]>
+    public class GetPayloadBodiesByHashV1Handler : IAsyncHandler<Keccak[], ExecutionPayloadBodyV1Result?[]>
     {
         private readonly IBlockTree _blockTree;
         private readonly ILogger _logger;
 
-        public GetPayloadBodiesV1Handler(IBlockTree blockTree, ILogManager logManager)
+        public GetPayloadBodiesByHashV1Handler(IBlockTree blockTree, ILogManager logManager)
         {
             _blockTree = blockTree;
             _logger = logManager.GetClassLogger();
         }
 
-        public Task<ResultWrapper<ExecutionPayloadBodyV1Result[]>> HandleAsync(Keccak[] blockHashes)
+        public Task<ResultWrapper<ExecutionPayloadBodyV1Result?[]>> HandleAsync(Keccak[] blockHashes)
         {
-            List<ExecutionPayloadBodyV1Result> payloadBodies = new(blockHashes.Length);
-            foreach (Keccak hash in blockHashes)
+            var payloadBodies = new ExecutionPayloadBodyV1Result?[blockHashes.Length];
+            for (int i = 0; i < blockHashes.Length; ++i)
             {
-                Block? block = _blockTree.FindBlock(hash);
-                if (block is not null)
-                {
-                    payloadBodies.Add(new ExecutionPayloadBodyV1Result(block.Transactions));
-                }
+                Block? block = _blockTree.FindBlock(blockHashes[i]);
+
+                payloadBodies[i] = block is not null ? new ExecutionPayloadBodyV1Result(block.Transactions) : null;
             }
 
-            return Task.FromResult(ResultWrapper<ExecutionPayloadBodyV1Result[]>.Success(payloadBodies.ToArray()));
+            return Task.FromResult(ResultWrapper<ExecutionPayloadBodyV1Result?[]>.Success(payloadBodies));
```

Comment:
Question - maybe even better would be to change result from array to IEnumerable<> and go with yield return? Would avoid even more allocations.

Reply:
Would play less nicely with Exception though, as this would be evaluated during serialization.

Reply:
One other idea to explore. With this branch: feature/json_dispose_result , we could use ArrayPoolList here. Just need to also dispose it if exception is thrown.